lib/bundler/definition.rb

```diff
@@ -105,8 +107,9 @@ def initialize(lockfile, dependencies, sources, unlock, ruby_version = nil, opti
 
       @gem_version_promoter = create_gem_version_promoter
 
-      current_platform = Bundler.rubygems.platforms.map {|p| generic(p) }.compact.last
-      add_platform(current_platform)
+      current_platform = Bundler.rubygems.platforms.last
+      add_platform(current_platform) if Bundler.settings[:specific_platform]
```

Comment:
we'll probably need to special-case around when in deployment mode and not add any platforms and count on the platforms check in https://github.com/bundler/bundler/pull/4895